packages/next-codemod/transforms/next-async-request-api/next-async-dynamic-api.ts

```diff
@@ -0,0 +1,430 @@
+              closestFunctionNode &&
+              isFunctionType(closestFunctionNode.type)
+            ) {
+              exportFunctionNode = closestFunctionNode
+            }
+          } else {
+            // Is normal async function
+            exportFunctionNode = closestFunctionNode
+          }
+
+          let canConvertToAsync = false
+          // check if current path is under the default export function
+          if (isFromExport) {
+            // if default export function is not async, convert it to async, and await the api call
+            if (!isCallAwaited) {
+              // If the scoped function is async function
+              if (
+                isFunctionType(exportFunctionNode.type) &&
+                exportFunctionNode.async === false
+              ) {
+                canConvertToAsync = true
+                exportFunctionNode.async = true
+              }
+
+              if (canConvertToAsync) {
+                j(path).replaceWith(
+                  j.awaitExpression(
+                    j.callExpression(j.identifier(asyncRequestApiName), [])
+                  )
+                )
+              }
+            }
+          } else {
+            // if parent is function function and it's a hook, which starts with 'use', wrap the api call with 'use()'
+            const parentFunction =
+              j(path).closest(j.FunctionDeclaration) ||
+              j(path).closest(j.FunctionExpression) ||
+              j(path).closest(j.ArrowFunctionExpression)
+
+            if (parentFunction.size() > 0) {
+              const parentFUnctionName = parentFunction.get().node.id?.name
+              const isParentFunctionHook = parentFUnctionName?.startsWith('use')
+              if (isParentFunctionHook) {
+                j(path).replaceWith(
+                  j.callExpression(j.identifier('use'), [
+                    j.callExpression(j.identifier(asyncRequestApiName), []),
+                  ])
+                )
+                needsReactUseImport = true
+              } else {
+                castTypesOrAddComment(
+                  j,
+                  path,
+                  asyncRequestApiName,
+                  root,
+                  filePath
+                )
+              }
+            } else {
+              castTypesOrAddComment(
+                j,
+                path,
+                asyncRequestApiName,
+                root,
+                filePath
+              )
+            }
+          }
+        }
+      })
+  }
+
+  const isClientComponent = isClientComponentAst(j, root)
+
+  // Only transform the valid calls in server or shared components
+  if (!isClientComponent) {
+    processAsyncApiCalls('cookies')
+    processAsyncApiCalls('headers')
+    processAsyncApiCalls('draftMode')
+
+    // Add import { use } from 'react' if needed and not already imported
+    if (needsReactUseImport) {
+      insertReactUseImport(root, j)
+    }
+  }
+
+  return root.toSource()
```

Comment:
You want to only `toSource` if something changed. Otherwise jscodeshift will touch a bunch of files it doesn't actually changed due to it's generator sometimes producing slightly different formatting. Front had 32 files changed that weren't actually affected.